execute/tokendata/observer_background.go

```diff
@@ -0,0 +1,431 @@
+		wg:                sync.WaitGroup{},
+		done:              doneChan,
+		observeTimeout:    observeTimeout,
+		reprocessInterval: 5 * time.Second,
+	}
+
+	o.startWorkers()
+	return o
+}
+
+// Observe fetches token data for the given messages that are already present in memory.
+// If token data are not in memory, it enqueues the messages for background processing.
+// Meaning that at least two calls to Observe are needed to get token data for a message.
+// If data are already in the queue for processing, it is a nop.
+func (o *backgroundObserver) Observe(
+	_ context.Context,
+	observations exectypes.MessageObservations,
+) (exectypes.TokenDataObservations, error) {
+	o.lggr.Debug("Observe called",
+		"observations", observations,
+		"cachedTokenData", o.cachedTokenData.size(),
+		"queuedMsgs", o.msgQueue.size(),
+	)
+
+	tokenDataResults := make(exectypes.TokenDataObservations)
+
+	// initialize the response with non-ready data since we have to respond with the same structure and
+	// number of elements as the input
+	for chainSel, seqNumToMsg := range observations {
+		tokenDataResults[chainSel] = make(map[cciptypes.SeqNum]exectypes.MessageTokenData)
+		for seqNum, msg := range seqNumToMsg {
+			tokenDataResults[chainSel][seqNum] = exectypes.MessageTokenData{
+				TokenData: make([]exectypes.TokenData, len(msg.TokenAmounts)),
+			}
+
+			for i := range msg.TokenAmounts {
+				tokenDataResults[chainSel][seqNum].TokenData[i] = exectypes.TokenData{
+					Ready:     false,
+					Supported: o.IsTokenSupported(chainSel, msg.TokenAmounts[i]),
+				}
+			}
+		}
+	}
+
+	// override with the cached data that are ready
+	for chainSel, seqNumToMsg := range observations {
+		for seqNum, msg := range seqNumToMsg {
+			tokenData, exists := o.cachedTokenData.get(msg.Header.MessageID)
+			if exists && !tokenData.SupportedAreReady() {
+				return nil, fmt.Errorf("internal error, cache contains not ready token data")
+			}
+
+			lggr := logger.With(o.lggr, "msgID", msg.Header.MessageID.String(),
+				"sourceChain", chainSel.String(), "seqNum", seqNum.String())
+
+			if exists {
+				// token data exist so include them in the results
+				lggr.Infow("token data found in cache")
+				tokenDataResults[chainSel][seqNum] = tokenData
+			} else {
+				// token data not in cache for this message, enqueue the message
+				if ok := o.msgQueue.enqueue(msg, 0); ok {
+					lggr.Infow("message added to the queue")
+				} else {
+					lggr.Infow("message already exists in the queue")
+				}
+			}
+		}
+	}
+
+	return tokenDataResults, nil
+}
+
+// IsTokenSupported simply forwards the call to the underlying observer.
+func (o *backgroundObserver) IsTokenSupported(
+	sourceChain cciptypes.ChainSelector, msgToken cciptypes.RampTokenAmount) bool {
+	return o.observer.IsTokenSupported(sourceChain, msgToken)
+}
+
+// Close stops the background goroutines and cleans up resources.
+func (o *backgroundObserver) Close() {
```

Comment:
Do we ever call that from the plugin? I guess we should call that during plugin shutdown